### PR TITLE
Provide contract choice metadata

### LINKED ISSUE #14329: Provide contract choice metadata as constants in contract's class
### What is the problem you want to solve?

The upcoming deprecated exercise by key command currently hard codes the choice name such as:

```java
...

  public static ExerciseByKeyCommand exerciseByKeyCreateManager(String key,
      CreateManager arg) {
    return new ExerciseByKeyCommand(Organization.TEMPLATE_ID, new Party(key), "CreateManager", arg.toValue());
  }

..
```

The choice name is hard coded and cannot be referenced / re-used.

If I create the command manually I would use:

```java
ExerciseCommand(Organization.TEMPLATE_ID, "someContractId", "CreateManager", args)
```

But again I have to know about the choice name and if the choice name changed in the future, compile time errors would not appear.

### What is the solution you would propose?

Could we get the choice names on a contract as constants (or some equivalent) so we can insert them as references?

It would also be helpful to have some contextual information about choices, such as is it a consuming or non-consuming, etc choice, so we can code around expected behaviour: example knowing it is a consuming choice can mean we remove it from the UI on submission (and re-add it if command eventually fails), vs non-consuming we would never remove it.
But would settle for just have the constants to start!

### PR DESCRIPTION
**Pull request by fayi-da**
Fixes #14329.

CHANGELOG_BEGIN
CHANGELOG_END

### CODE REVIEW

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractCompanion.java [hunk 1]
```diff
@@ -44,7 +46,8 @@ protected ContractCompanion(
       String templateClassName,
       Identifier templateId,
       Function<String, Id> newContractId,
-      Function<DamlRecord, Data> fromValue) {
+      Function<DamlRecord, Data> fromValue,
+      List<ChoiceMetadata<Data, ?, ?>> choices) {
```
Comment (S11001001) [inline language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractCompanion.java:50]: Here's what I suggest we do with this thing:

1. add a `Marker` type parameter to `ContractTypeCompanion`. For `InterfaceCompanion` it should be `I`, for `ContractCompanion` it should be `Data`
2. Pass `choices` along to `super` (substituting `Marker` for `Data` in the type)
3. transform it to the instance variable `public final Map<String, ChoiceMetadata<Marker, ?, ?>> choices` (see `ContractDecoder` for an example that makes it safe and immutable)

That would be reasonable for this PR; for interfaces we can either include them in this PR or do a separate PR to supply them.

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractTypeCompanion.java [hunk 2]
```diff
@@ -5,12 +5,17 @@
...
 import com.daml.ledger.javaapi.data.Identifier;
 
+import java.util.List;
+
 /** The commonality between {@link ContractCompanion} and {@link InterfaceCompanion}. */
 public abstract class ContractTypeCompanion {
   /** The full template ID of the template or interface that defined this companion. */
   public final Identifier TEMPLATE_ID;
+  //TODO: Need to figure out the right type params here
+  private final List<ChoiceMetadata<?, ?, ?>> choices;
```
Comment (fayi-da) [inline language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractTypeCompanion.java:15]: @S11001001 I was about to ask about this and saw your comment from yesterday. I think adding that `Marker` class will solve my problem

Comment (S11001001) [reply language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractTypeCompanion.java:15]: We're going to change it again in #15125, so might as well get started now.

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/InterfaceClass.scala [hunk 3]
```diff
@@ -29,6 +32,16 @@ object InterfaceClass extends StrictLogging {
         .classBuilder(interfaceName)
         .addModifiers(Modifier.FINAL, Modifier.PUBLIC)
         .addField(generateTemplateIdField(packageId, interfaceId))
+        .addFields(
+          TemplateClass
+            .generateChoicesMetadata(
+              interfaceName,
+              packagePrefixes,
+              interface.choices,
+              withPrefixes = false, // TODO: remove in #15154
```
Comment (fayi-da) [inline language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/InterfaceClass.scala:41]: Removing this in the [next ticket ](https://github.com/digital-asset/daml/issues/15154)

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractTypeCompanion.java [hunk 4]
```diff
@@ -5,18 +5,22 @@
 
 import com.daml.ledger.javaapi.data.Identifier;
 
+import java.util.List;
+
 /** The commonality between {@link ContractCompanion} and {@link InterfaceCompanion}. */
 public abstract class ContractTypeCompanion<Maker, Data> {
```
Comment (ray-roestenburg-da) [inline language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractTypeCompanion.java:11]: Not clear what this is, let's discuss with @S11001001

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractTypeCompanion.java [hunk 5]
```diff
@@ -5,18 +5,22 @@
...
   /** The full template ID of the template or interface that defined this companion. */
   public final Identifier TEMPLATE_ID;
+  private final List<ChoiceMetadata<Maker, ?, ?>> choices;
```
Comment (S11001001) [inline language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractTypeCompanion.java:14]: Which can be made in the constructor here.

```suggestion
  private final Map<String, ChoiceMetadata<Marker, ?, ?>> choices;
```

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractTypeCompanion.java (hunk 4 above)
Comment (S11001001) [reply language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractTypeCompanion.java:11]: @ray-roestenburg-da `ContractCompanion` represents templates, `InterfaceCompanion` interfaces, and the union of the two...

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ChoiceMetadata.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
+package com.daml.ledger.javaapi.data.codegen;
+
+import com.daml.ledger.javaapi.data.Value;
+
+import java.util.function.Function;
+
+public final class ChoiceMetadata<Tpl, ArgType, ResType> {
+    public final String name;
```
Comment (S11001001) [inline language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ChoiceMetadata.java:8]: These are intended as part of the user API and so should be Javadoc'd.

**Review by S11001001 — COMMENTED** [review]
Run `./fmt.sh --diff` in the root to fix the Java format.

Additionally, sample `CHOICE_*` fields should be documented in `docs/source/app-dev/bindings-java/codegen.rst`

* [x] near the definition of `COMPANION` for templates
* [ ] near the definition of field `INTERFACE` for interfaces

The details of the initializer can be elided; what matters is that it is clear that such fields are defined, what their types are, and users can refer to Javadoc for further information.

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractTypeCompanion.java (hunk 4 above)
Comment (S11001001) [reply language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractTypeCompanion.java:11]: (from other discussion) For an interface, the type argument to `ContractId` (`Marker`) is different from the payload type i.e. interface view (`Data`); for a template they are the same.

**Review by S11001001 — COMMENTED** [review]
Additionally,

* [ ] test that you can use a `CHOICE_*` field to retrieve the name of a choice in codegen output
* [ ] test that the `choices` maps' keysets in a `COMPANION` and an `INTERFACE` correspond to expected sets of choice names

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractTypeCompanion.java (hunk 5 above)
Comment (ray-roestenburg-da) [reply language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractTypeCompanion.java]: I would guess the choice name.

**Review by carrielaben-da — COMMENTED** [review]
Docs portion looks good.

File: docs/source/app-dev/bindings-java/codegen.rst [hunk 7]
```diff
@@ -182,9 +182,12 @@ A file is generated that defines five Java classes and an interface:
 
     public static final Identifier TEMPLATE_ID = new Identifier("some-package-id", "Com.Acme.Templates", "Bar");
 
+    public static final ChoiceMetadata<Bar, Archive, Unit> CHOICE_Archive =
+      ChoiceMetadata.create(/* ... */);
+
```
[the remark this answers is not in the document]
Comment (S11001001) [reply docs/source/app-dev/bindings-java/codegen.rst:187]: If you like, you can check the docs preview and make sure these highlights point to the classes/interfaces that the docs are referring to, but I'm pretty sure I got the arithmetic right.

**Review by ray-roestenburg-da — APPROVED** [review]

**Pull request merged**